device-api/src/topology/mod.rs

```diff
@@ -0,0 +1,349 @@
+#![allow(warnings)]
+use std::collections::BTreeMap;
+
+use crate::topology::hwloc::{Hwloc, HwlocTopology};
+use crate::topology::hwloc_binding::*;
+use crate::topology::LinkType::*;
+use crate::{Device, DeviceResult};
+
+mod build;
+mod hwloc;
+mod hwloc_binding;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum LinkType {
+    // LinkTypeUnknown unknown
+    LinkTypeUnknown = 0,
+    // LinkTypeInterconnect two devices are connected across different cpus through interconnect.
+    LinkTypeInterconnect = 10,
+    // LinkTypeCPU two devices are connected under the same cpu, it may mean:
+    // devices are directly attached to the cpu pcie lane without PCIE switch.
+    // devices are attached to different PCIE switches under the same cpu.
+    LinkTypeCPU = 20,
+    // LinkTypeHostBridge two devices are connected under the same PCIE host bridge.
+    // Note that this does not guarantee devices are attached to the same PCIE switch.
+    // More switches could exist under the host bridge switch.
+    LinkTypeHostBridge = 30,
+
+    // NOTE(@bg): Score 40 and 50 is reserved for LinkTypeMultiSwitch and LinkTypeSingleSwitch.
+    // NOTE(@bg): Score 60 is reserved for LinkTypeBoard
+
+    // LinkTypeSoc two devices are on the same Soc chip.
+    LinkTypeSoc = 70,
+}
+
+impl LinkType {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            LinkTypeUnknown => "LinkTypeUnknown",
+            LinkTypeInterconnect => "LinkTypeInterconnect",
+            LinkTypeCPU => "LinkTypeCPU",
+            LinkTypeHostBridge => "LinkTypeHostBridge",
+            LinkTypeSoc => "LinkTypeSoc",
+        }
+    }
+}
+
+pub trait HardwareTopologyHint {
+    fn get_hw_topology_hint(&self) -> String;
```

Comment:
at this moment we only use bdf for hint but we don't know what will be used along with bdf.
So in my humble opinion, it would be nice if we keep arbitrary naming for that.

Reply:
I agree that other stuff could be added for topology hints. However, even in that case, I think the method should be added to the `HardwareTopologyHint` trait, not the interface and implementation of the function returning the bdf string should be changed.

Reply:
@libc-furiosa 
The role of this Trait is to embrace something other than bdf in the future. 
The purpose of this trait is to abstract heterogeneous data input into an interface.
```rust
get_hw_topology_hint(&self) -> String;
```

If we add specific methods for the trait, it will be inconvenient for both the provider and the consumer.

From the provider's perspective, if trait says multiple methods, then structs such as Device or Nic(for example) must implement all of them. 

From the consumer's perspective(such as algorithm to select devices), it is difficult to determine which of the multiple methods to use.  Calling all method does not make sense.

Therefore, adding methods to this trait is not appropriate way, we should avoid that way.

I think returning a Struct containing the data with an enum specifying the type of data is okay, instead of returning a String.
```rust
pub enum TopologyHintType {
    BDF = 0,
    // reserved for future usage 
}

pub struct TopologyHint {
 pub type: TopologyHintType,
 pub hint: String
}

pub trait HardwareTopologyHint {
    fn get_hw_topology_hint(&self) -> TopologyHint;
}
```

Reply:
resolved
f9fc416c41750c649292bc97103a41607c44f1fd